src/Components/Web/src/Virtualization/Virtualize.cs

```diff
@@ -0,0 +1,304 @@
+        }
+
+        void IVirtualizeJsCallbacks.OnAfterSpacerVisible(float spacerSize, float containerSize)
+        {
+            CalcualteItemDistribution(spacerSize, containerSize, out var itemsAfter, out var visibleItemCapacity);
+
+            var itemsBefore = Math.Max(0, _itemCount - itemsAfter - visibleItemCapacity);
+
+            UpdateItemDistribution(itemsBefore, visibleItemCapacity);
+        }
+
+        private void CalcualteItemDistribution(float spacerSize, float containerSize, out int itemsInSpacer, out int visibleItemCapacity)
+        {
+            itemsInSpacer = Math.Max(0, (int)Math.Floor(spacerSize / ItemSize) - 1);
+            visibleItemCapacity = (int)Math.Ceiling(containerSize / ItemSize) + 2;
+        }
+
+        private void UpdateItemDistribution(int itemsBefore, int visibleItemCapacity)
+        {
+            if (itemsBefore != _itemsBefore || visibleItemCapacity != _visibleItemCapacity)
+            {
+                _itemsBefore = itemsBefore;
+                _visibleItemCapacity = visibleItemCapacity;
+                var refreshTask = RefreshDataAsync();
+
+                if (!refreshTask.IsCompleted)
+                {
+                    StateHasChanged();
+                }
+            }
+        }
+
+        private async Task RefreshDataAsync()
+        {
+            _refreshCts?.Cancel();
+            _refreshCts = new CancellationTokenSource();
+
+            var cancellationToken = _refreshCts.Token;
+            var request = new ItemsProviderRequest(_itemsBefore, _visibleItemCapacity, cancellationToken);
+
+            try
+            {
+                var result = await _itemsProvider(request);
+
+                // Only apply result if the task was not canceled.
+                if (!cancellationToken.IsCancellationRequested)
+                {
+                    _itemCount = result.TotalItemCount;
+                    _loadedItems = result.Items;
+                    _loadedItemsStartIndex = request.StartIndex;
+
+                    StateHasChanged();
+                }
+            }
+            catch (Exception e)
+            {
+                if (e is OperationCanceledException oce && oce.CancellationToken == cancellationToken)
+                {
+                    // No-op; we canceled the operation, so it's fine to suppress this exception.
+                }
+                else
+                {
+                    // Cache this exception so the renderer can throw it.
+                    _refreshException = e;
+
+                    // Re-render the component to throw the exception.
+                    StateHasChanged();
+                }
+            }
+        }
+
+        private ValueTask<ItemsProviderResult<TItem>> DefaultItemsProvider(ItemsProviderRequest request)
+        {
+            return ValueTask.FromResult(new ItemsProviderResult<TItem>(
+                Items!.Skip(request.StartIndex).Take(request.Count),
```

Comment:
I'm trying to figure out whether `Enumerable.Skip` has an optimization for indexable types where it goes directly to the requested index, or whether it actually walks the enumerator N times. In [these sources it seems to walk the enumerator](https://referencesource.microsoft.com/#system.core/system/linq/Enumerable.cs,83ec6a20321060a1), and [this discussion](https://stackoverflow.com/questions/20002975/performance-of-skip-and-similar-functions-like-take) implies that's what it does too. However I set up a little microbenchmark and was unable to observe any extra perf cost from walking a long collection, even with a billion elements.

If it turns out that walking an `ICollection`'s enumerator to implement `Skip` is an O(N) operation that becomes problematic for large N on WebAssembly, we could consider a future enhancement where we implement our own CollectionEnumerator that knows how to skip in O(1).

Maybe you or someone else reading this knows better about the internals of `Skip` and what it ends up costing. I'm not suggesting we need to optimize this right now!